### PR TITLE
allow test to run even if VS150COMTOOLS is not defined 

### PR DESCRIPTION
**Pull request by smoothdeveloper**
we can define it ourselves

### CODE REVIEW

File: vsintegration/tests/Salsa/VsMocks.fs [hunk 1]
```diff
@@ -1639,8 +1639,19 @@ module internal VsActual =
         if String.IsNullOrEmpty vsvar then failwith "VS140COMNTOOLS environment variable was not found."
 #endif
 #if VS_VERSION_DEV15
-        let vsvar = System.Environment.GetEnvironmentVariable("VS150COMNTOOLS")
-        if String.IsNullOrEmpty vsvar then failwith "VS150COMNTOOLS environment variable was not found."
+        let vsvar = 
+            let vsvar = System.Environment.GetEnvironmentVariable("VS150COMNTOOLS")
+            if String.IsNullOrEmpty vsvar then
+                let maybeDirectory = 
+                    [ @"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\Tools\" ]
```
Comment (vasily-kirichenko) [inline vsintegration/tests/Salsa/VsMocks.fs]: Why "Community" only? What about Professional and Enterprise? Maybe there is an env var which stores VS 2017 root dir?

Comment (smoothdeveloper) [inline vsintegration/tests/Salsa/VsMocks.fs]: I don't know the folders, if you have them please let me know and I'll add them.

Comment (vasily-kirichenko) [inline vsintegration/tests/Salsa/VsMocks.fs]: `c:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\`
`c:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\`

Comment (smoothdeveloper) [inline vsintegration/tests/Salsa/VsMocks.fs]: Thanks!

**Comment by smoothdeveloper** [comment]
@saul I don't see the InstallDir in registry, I'd say if it break on someone, that someone can then fix it, right now it breaks on everyones and it fixes it for 99.99% of them.

**Comment by saul** [comment]
@smoothdeveloper are you looking under Wow6432Node? This value always exists.

**Comment by smoothdeveloper** [comment]
@saul yes, it doesn't exist anymore for VS2017, at least on my machine.

image

I'm making that change so it is easy for most people to run tests from VS itself (without having to run it from a special shell), it can be improved further but this is already an improvement I think.

**Comment by saul** [comment]
Urgh they got rid of it in 2017 😒

HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\VisualStudio\SxS\VS7 use the 15.0 value.

**Comment by smoothdeveloper** [comment]
@saul / anyone, can someone confirm what is the situation in registry when several VS2017 are installed (I believe you can have them side by side)

image

I'm ok adding registry to the list of things we check, but I'd not like it to break because we'd have removed the hardcoded folders and that it doesn't look the same when there are several editions installed.

**Comment by saul** [comment]
Good question - what do we use the environment variable for out of interest? Is it just for the F# compiler - if so it doesn't really matter which edition we use

**Comment by smoothdeveloper** [comment]
@saul, looking a bit at usage (which I didn't do when I was just fixing that to get tests running in my VS), it is just bellow: it prepares an `AssemblyCatalog` from "IDE\CommonExtensions\Microsoft\Editor" folder.

It looks like MEF components gets registered this way, and for the purpose of unit tests, I don't think the different edition would have an impact.

We can still add better resolution with registry, I've made some more change to make that area of the code a bit better.

**Comment by KevinRansom** [comment]
Start a VS command window for the version of VS you want it to run against and it will set the variables you need.  Replacing that script with a bunch of code is not really that interesting, and likely to be a maintenance bug farm.

So ... I am inclined to close this PR, there are plenty of places where we have code and scripts that adapt to the lack of this configuration, and they always have introduced unreliability and confusion and make it difficult to reason about or discover what happens.

What is the scenario you have in mind where the tests are run in the absence of running that script.

Kevin

**Comment by smoothdeveloper** [comment]
@KevinRansom the scenario I'm using it for:

* click VS2017
* open VisualFSharp.sln
* run any of the salsa tests from "Test Explorer" (many people like to run unit test from VS, either with Resharper or NUnit test adapter)

The tests don't rely on all the environment variables, they just need the location to the "IDE\CommonExtensions\Microsoft\Editor" folder.

It is ok to close it of course, but the concern of having tests run on most machines (without having to restart VS because it wasn't started from specific command prompt).

Generally to make the work in this repository look more like normal workflow (click build.cmd, open VS and get ready to work) is important concern.

@KevinRansom maybe ask the VS team to make that devenv will set the same stuff as the command line, that might be a sensible thing for future VS releases?

**Comment by KevinRansom** [comment]
@smooth developer I haven't had a chance to see if this could help, but perhaps it might.  

https://msdn.microsoft.com/en-us/library/jj635153.aspx

**Comment by smoothdeveloper** [comment]
@KevinRansom this is something users of VS test runner can use, but it won't work in third part test runner (like Resharper, which is way superior).

I understand the concern about replicating `C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\Tools\VsDevCmd.bat` but here we just want VS folder, only enabling this via environment variable, or requiring the developer to update a .runsettings hurts the workflow for the common developer.

**Comment by KevinRansom** [comment]
@smoothdeveloper 

Almost all of the files in that location are available in nuget packages.  Nuget seems like the appropriate long term solution, also it would also remove another dependency of our build on the install location of VS ... that would be an excellent outcome ...

````
Microsoft.VisualStudio.CoreUtility.dll                      ---- https://www.nuget.org/packages/Microsoft.VisualStudio.CoreUtility/
Microsoft.VisualStudio.Editor.dll                           ---- https://www.nuget.org/packages/Microsoft.VisualStudio.Editor/
Microsoft.VisualStudio.Language.StandardClassification.dll  ---- https://www.nuget.org/packages/Microsoft.VisualStudio.Text.Data/
Microsoft.VisualStudio.Text.Data.dll                        ---- https://www.nuget.org/packages/Microsoft.VisualStudio.Text.Data/
Microsoft.VisualStudio.Text.Logic.dll                       ---- https://www.nuget.org/packages/Microsoft.VisualStudio.Text.Logic
Microsoft.VisualStudio.Text.UI.dll                          ---- https://www.nuget.org/packages/Microsoft.VisualStudio.Text.UI
Microsoft.VisualStudio.Text.UI.Wpf.dll                      ---- https://www.nuget.org/packages/Microsoft.VisualStudio.Text.UI.Wpf


Microsoft.VisualStudio.Editor.Implementation.dll            ---- 
Microsoft.VisualStudio.Platform.VSEditor.dll                ---- https://dotnet.myget.org/feed/roslyn-tools/package/nuget/Microsoft.VisualStudio.Platform.VSEditor
````

We may not require : Microsoft.VisualStudio.Editor.Implementation.dll and Microsoft.VisualStudio.Platform.VSEditor.dll.  If we do we can try to get them both added to nuget.  I'm sure we can get Microsoft.VisualStudio.Editor.Implementation.dll  added to myget?

I realize it is more work ... but releasing ourselves from an install location dependency seems like a win, what do you think?

Kevin

**Comment by cloudRoutine** [comment]
@KevinRansom I'd definitely be  a big fan of it, VSSDK version conflicts and directory changes and all the other finicky maddening configuration twiddles and tweaks are exasperating. The more consistency and stability when it comes to these dependencies the better 😁

**Comment by smoothdeveloper** [comment]
Good idea @KevinRansom, I'll look to get it work this way.

**Comment by smoothdeveloper** [comment]
@KevinRansom I'm seeing there are pending work on the build in several PRs, which might impact the packages.config stuff / location of packages.

Can we get this PR merged and postpone further change to leverage nuget packages or should I close this?

**Comment by KevinRansom** [comment]
@smoothdeveloper I would prefer a nuget based solution to this issue.  we can close this and make an issue suggesting that we move the dll dependencies to a nugget package.

**Comment by cloudRoutine** [comment]
@KevinRansom that's what I did in https://github.com/Microsoft/visualfsharp/pull/2698

**Comment by KevinRansom** [comment]
I think we need to fix this a different way.

**Pull request closed without merging**